rust/tableau_summary/src/twb/raw/datasource/connection.rs

```diff
@@ -0,0 +1,489 @@
+            union_table: Table::from(n),
+            tables,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, PartialEq, Clone, Debug)]
+pub struct Clause {
+    pub clause_type: String,
+    pub expression: Expression,
+}
+
+#[derive(Serialize, Deserialize, Default, PartialEq, Clone, Debug)]
+pub struct Expression {
+    pub op: String,
+    pub expressions: Vec<Expression>,
+}
+
+impl<'a, 'b> From<Node<'a, 'b>> for Expression {
+    fn from(n: Node) -> Self {
+        if n.get_tag() != "expression" {
+            info!("trying to convert a ({}) to an expression", n.get_tag());
+            return Self::default();
+        }
+        let expressions = n.find_tagged_children("expression")
+            .into_iter()
+            .map(Expression::from)
+            .collect();
+
+        Self {
+            op: n.get_attr("op"),
+            expressions,
+        }
+    }
+}
+
+
+#[derive(Serialize, Deserialize, Default, PartialEq, Clone, Debug)]
+pub struct TableColumn {
+    pub name: String,
+    pub datatype: String,
+    pub ordinal: usize,
+}
+
+impl<'a, 'b> From<Node<'a, 'b>> for TableColumn {
+    fn from(n: Node) -> Self {
+        if n.get_tag() != "column" {
+            info!("trying to convert a ({}) to a column", n.get_tag());
+            return Self::default();
+        }
+        let ordinal = n.get_maybe_attr("ordinal")
+            .map(|s|s.parse()
+                .log_error("ordinal not a number")
+                .unwrap_or_default())
+            .unwrap_or_default();
+        Self {
+            name: n.get_attr(NAME_KEY),
+            datatype: n.get_attr("datatype"),
+            ordinal,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, PartialEq, Clone, Debug)]
+pub struct ColMapping {
+    pub cols: HashMap<String, (String, String)>,
+}
+
+impl<'a, 'b> From<Node<'a, 'b>> for ColMapping {
+    fn from(n: Node) -> Self {
+        if n.get_tag() != "cols" {
+            info!("trying to convert a ({}) to cols", n.get_tag());
+            return Self::default();
+        }
+        let cols = n.find_tagged_children("map")
+            .into_iter()
+            .flat_map(map_node_to_kv)
+            .collect();
+        Self {
+            cols,
+        }
+    }
+}
+
+fn map_node_to_kv(n: Node) -> Option<(String, (String, String))> {
+    let key = n.get_attr("key");
+    let val = n.get_attr("value");
+    let val_parts = parse_identifiers(&val);
+    let val_tuple = if let Some((x, y)) = val_parts.into_iter().collect_tuple() {
+        (x, y)
+    } else {
+        info!("found col value: {val} that didn't parse into 2 parts");
+        return None;
+    };
+    Some((key, val_tuple))
+}
+
+// Given a string like `[foo].[bar]` return ["foo", "bar"]
+// TODO: see what tds / twb files do when the table or column has a `.` in the name
+//       currently, we just assume it is not escaped or anything
+//       (e.g. "[foo.1].[bar.txt]" -> ["foo.1", "bar.txt"])
```

Comment:
sadly this sounds like a good regex problem.
starting point: https://chat.openai.com/share/8b8a54e3-58be-4585-b848-7cde9a964363